### DISCUSSION TITLE
Support Forge/Fabric Mods

### DISCUSSION THREAD

**Discussion by dest1n1s** [valence-rs/valence#669] [Q&A]
Hello everyone! I want to ask if `valence` plans to support mods in the future, or if it just serves a vanilla Minecraft server?

**Comment by Earthcomputer** [comment]
Forge/Fabric mods are written in Java, this is a Rust project. Any functionality that a mod adds will have to be reimplemented in Rust. Although this is a server framework, not a full server reimplementation, so you would have to implement the base game too

**Comment by dyc3** [comment]
Valence is not a minecraft server, its a server framework. You have to build all the functionality you want yourself. Of course, any client-side only mods will work just fine.

That being said, if you are wondering if you could write a server using valence, and have some custom blocks or items that come from a mod, ideally that should be possible but it currently is not. Particularly, the set of items and blocks that valence knows about are generated from vanilla minecraft, and we don't currently expose an api surface to register new blocks or items. 

Realistically though, most use cases won't benefit from this approach. The Minecraft client is quite dyanamic and configurable enough from the server side (see: game servers like hypixel skyblock).

**Comment by dest1n1s** [comment]
Thank you all for your helpful responses!

I'm actually looking to rewrite some Minecraft tech mods in Rust (e.g. the ones from [GT:NH](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack)), but I'm not sure where to start. Do I just need to handle block/item/entity textures and GUIs on the client side, while putting all the block update logic on the server?

**Reply by dyc3** [reply]
Pretty much, but that's simplifying it quite a bit. 

Like I said before, we don't expose any APIs yet that would allow users to add custom blocks. You would need to maintain your own fork of valence (really not recommended), or we would need to refactor items and blocks as registries instead of enums (at minimum). I have no idea what else you would need, nobody has tried this before. PRs are welcome, of course.

**Reply by dest1n1s** [reply]
> Pretty much, but that's simplifying it quite a bit.
> 
> Like I said before, we don't expose any APIs yet that would allow users to add custom blocks. You would need to maintain your own fork of valence (really not recommended), or we would need to refactor items and blocks as registries instead of enums (at minimum). I have no idea what else you would need, nobody has tried this before. PRs are welcome, of course.

Thanks! I'll spend some time evaluating the difficulties to implement all these stuffs.
